quickjs.c

```diff
@@ -9198,6 +9248,62 @@ static int js_shape_prepare_update(JSContext *ctx, JSObject *p,
     return 0;
 }
 
+int js_shape_delete_watchpoints(JSRuntime *rt, JSShape *shape, intptr_t target) {
+    struct list_head *el, *el1;
+    if (unlikely(!shape || !shape->watchpoint))
+        goto end;
+    list_for_each_safe(el, el1, shape->watchpoint) {
+        JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
+        if (o->delete_callback)
+            if (!o->delete_callback(rt, (intptr_t)o->ref, o->offset, o->atom, target)) {
+                list_del(el);
+                js_free_rt(rt, o);
+                break;
+            }
+    }
+end:
+    return 0;
+}
+
+int js_shape_free_watchpoints(JSRuntime *rt, JSShape *shape) {
+    struct list_head *el, *el1;
+    if (unlikely(!shape || !shape->watchpoint))
+        goto end;
+    list_for_each_safe(el, el1, shape->watchpoint) {
+        JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
+        if (o->free_callback)
+            o->free_callback(rt, (intptr_t)o->ref, o->offset, o->atom);
+        list_del(el);
+        js_free_rt(rt, o);
+    }
+    list_empty(shape->watchpoint);
+    js_free_rt(rt, shape->watchpoint);
+end:
+    return 0;
+}
+
+JSInlineCacheWatchpoint *js_shape_create_watchpoint(JSRuntime *rt, JSShape *shape, intptr_t ptr, uint8_t offset, JSAtom atom,
+                                                    watchpoint_delete_callback *delete_callback,
+                                                    watchpoint_free_callback *free_callback)
+{
+    JSInlineCacheWatchpoint *o;
+    o = (JSInlineCacheWatchpoint *)js_malloc_rt(rt, sizeof(JSInlineCacheWatchpoint));
+    if (unlikely(!o))
+        return NULL;
+    o->ref = ptr;
+    o->atom = atom;
+    o->offset = offset;
+    o->delete_callback = delete_callback;
+    o->free_callback = free_callback;
+    if (!shape->watchpoint) {
+        shape->watchpoint = (struct list_head *)js_malloc_rt(rt, sizeof(struct list_head));
```

Comment:
Ditto, and consider inlining it in JSShape (or add a NULL check here.)

I'm guessing you don't want to enlarge JSShape too much (laudable goal) but it's only one extra word vs. a pointer. I'd say that's okay.